scipy/stats/mstats_extras.py

```diff
@@ -161,13 +161,12 @@ def _hdsd_1D(data, prob):
 
         vv = np.arange(n) / float(n-1)
         betacdf = beta.cdf
-
+        
         for (i,p) in enumerate(prob):
             _w = betacdf(vv, (n+1)*p, (n+1)*(1-p))
             w = _w[1:] - _w[:-1]
-            mx_ = np.fromiter([np.dot(w,xsorted[np.r_[list(range(0,k)),
-                                                      list(range(k+1,n))].astype(int_)])
-                                  for k in range(n)], dtype=float_)
+            mx_ = np.fromiter([np.dot(w, np.delete(xsorted, k))
```

Comment:
I don't know the shapes of the arguments here but first using `@` for the product is more readable and also first multiplying and then picking might be quicker. Do you have any info on those?

Reply:
Not sure what you mean, sorry. Could you write it in code or explain it with an example?

Reply:
Picking up a subarray is more expensive in the memory if the array is 2D. Picking up subsequence from 1D array is cheaper.

Reply:
Sorry, still not sure what you mean by "picking up a subarray" or what you are comparing to what. Can you just tell me (with code) the specific step within that code snippet that you think it's less efficient?

Also, have you looked at the test code I have provided, showing substantial speed up? Could you modify my test code in the way you think it'd be slower?

Reply:
FYI. `w` is a 1D vector, as you can see from the code

```
n = len(xsorted)
...
vv = np.arange(n) / float(n-1)
...
            _w = betacdf(vv, (n+1)*p, (n+1)*(1-p))
            w = _w[1:] - _w[:-1]
```

Reply:
I am talking about the delete part here so everytime we skip one element in xsorted right? That creates a new array each time. I don't understand the code here so I can give you an example, if xsorted is 2D then that is slower than removing element after the multiplication because the result will be 1D. If both are 1d I don't know what the dot product is doing there.

Reply:
Ah, OK, I think you are having trouble understanding the algorithm. The dot product is just part of the algorithm. And I think you have missed the line

```
xsorted = np.sort(data.compressed())
```

Thus, `xsorted` is going to be 1D. You have some Harrell-Davis weights `w`. You have one fewer weight than data points. For the sake of simplicity, let's make up a small example.

```
w = [0.1, 0.2, 0.3, 0.2, 0.1]
```

Then, you have the sorted data `xsorted`. It's a 1D vector, as seen above, e.g.

```
xsorted = [1, 3, 3, 5, 7, 10]
```

So, what you need to do in that part of the algorithm is

```
np.dot([0.1, 0.2, 0.3, 0.2, 0.1], [   3, 3, 5, 7, 10])
np.dot([0.1, 0.2, 0.3, 0.2, 0.1], [1,    3, 5, 7, 10])
np.dot([0.1, 0.2, 0.3, 0.2, 0.1], [1, 3,    5, 7, 10])
np.dot([0.1, 0.2, 0.3, 0.2, 0.1], [1, 3, 3,    7, 10])
np.dot([0.1, 0.2, 0.3, 0.2, 0.1], [1, 3, 3, 5,    10])
np.dot([0.1, 0.2, 0.3, 0.2, 0.1], [1, 3, 3, 5, 7    ])
```

Each dot product of two vectors gives you a scalar as a result. Of course, we could also build a matrix 

```
x_leave_one_out = 
[[   3, 3, 5, 7, 10],
 [1,    3, 5, 7, 10],
 [1, 3,    5, 7, 10],
 [1, 3, 3,    7, 10],
 [1, 3, 3, 5,    10],
 [1, 3, 3, 5, 7    ]]
```
and then do some kind of `x_leave_one_out * w.transpose()`. This avoids the loop, but the problem is that the matrix has size N^2, and this is impractical for long data vectors.

Reply:
Ah yes, very nice explanation. I got it now. It doesn't look like we have much options then indeed.